pkg/util/process/collector.go

```diff
@@ -0,0 +1,133 @@
+package process
+
+import (
+	"bufio"
+	"bytes"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	// DefaultProcMountPoint is the common mount point of the proc filesystem.
+	DefaultProcMountPoint = "/proc"
+)
+
+type processCollector struct {
+	pid            int
+	procMountPoint string
+	supported      bool
+
+	// Metrics.
+	currMaps *prometheus.Desc
+	maxMaps  *prometheus.Desc
+}
+
+// NewProcessCollector makes a new custom process collector used to collect process metrics the
+// default instrumentation doesn't support.
+func NewProcessCollector() prometheus.Collector {
+	return newProcessCollector(os.Getpid(), DefaultProcMountPoint)
+}
+
+func newProcessCollector(pid int, procMountPoint string) prometheus.Collector {
+	c := &processCollector{
+		pid:            pid,
+		procMountPoint: procMountPoint,
+		supported:      isSupported(procMountPoint),
+		currMaps: prometheus.NewDesc(
+			"process_memory_map_areas",
+			"Number of memory map areas allocated by the process.",
+			nil, nil,
+		),
+		maxMaps: prometheus.NewDesc(
+			"process_memory_map_areas_limit",
+			"Maximum number of memory map ares the process can allocate.",
+			nil, nil,
+		),
+	}
+
+	return c
+}
+
+// Describe returns all descriptions of the collector.
+func (c *processCollector) Describe(ch chan<- *prometheus.Desc) {
+	if !c.supported {
+		return
+	}
+
+	ch <- c.currMaps
+	ch <- c.maxMaps
+}
+
+// Collect returns the current state of all metrics of the collector.
+func (c *processCollector) Collect(ch chan<- prometheus.Metric) {
+	if !c.supported {
+		return
+	}
+
+	if value, err := c.getMapsCount(); err == nil {
+		ch <- prometheus.MustNewConstMetric(c.currMaps, prometheus.GaugeValue, value)
+	}
+
+	if value, err := c.getMapsCountLimit(); err == nil {
+		ch <- prometheus.MustNewConstMetric(c.maxMaps, prometheus.GaugeValue, value)
+	}
+}
+
+// getMapsCount returns the number of memory map ares the process has allocated.
+func (c *processCollector) getMapsCount() (float64, error) {
+	file, err := os.Open(processMapsPath(c.procMountPoint, c.pid))
+	if err != nil {
+		return 0, err
+	}
+	defer file.Close()
+
+	count := 0
+	for scan := bufio.NewScanner(file); scan.Scan(); {
+		count++
+	}
+
+	return float64(count), nil
+}
+
+// getMapsCountLimit returns the maximum of memory map ares the process can allocate.
+func (c *processCollector) getMapsCountLimit() (float64, error) {
+	file, err := os.Open(vmMapsLimitPath(c.procMountPoint))
+	if err != nil {
+		return 0, err
+	}
+	defer file.Close()
+
+	content, err := ioutil.ReadAll(file)
+	if err != nil {
+		return 0, err
+	}
+
+	content = bytes.TrimSpace(content)
+
+	// A base value of zero makes ParseInt infer the correct base using the
+	// string's prefix, if any.
+	const base = 0
+	value, err := strconv.ParseInt(string(content), base, 64)
```

Comment:
Base=0 seems wrong. If value starts with zero, it will be treated as 8-based. I don't think that's correct.

Reply:
I picked it from`procfs` implementation. I trust them. I would keep it.

Reply:
I don't think that is a correct decision, but let's commit to it and see how it works.